### ISSUE TITLE
SampleParticle time offset

### ISSUE THREAD

**Issue by apatlpo** [OceanParcels/parcels#1122]
When using a sample particle, it seems the output sampled field is offset by one timestep compared to the input field.
Is this expected behavior?

See the following example:

```
from parcels import FieldSet, ParticleSet, JITParticle, Variable

class SampleParticle(JITParticle):
    p = Variable('p', dtype=np.float32)

def SampleP(particle, fieldset, time):
    particle.p = fieldset.P[time, particle.depth, particle.lat, particle.lon]

dims = [5, 4, 2]
dx, dy = 1./dims[0], 1./dims[1]
dimensions = {'lon': np.linspace(0., 1., dims[0], dtype=np.float32),
              'lat': np.linspace(0., 1., dims[1], dtype=np.float32),
              'time': np.arange(dims[2], dtype=np.float32),
             }
data = {'U': np.zeros(dims, dtype=np.float32),
        'V': np.zeros(dims, dtype=np.float32),
        'P': np.ones((dims[0], dims[1]))[:,:,None]*np.arange(dims[2])[None, None, :]        
       }
fieldset = FieldSet.from_data(data, dimensions, mesh='flat', transpose=True)

xv, yv = np.meshgrid(np.linspace(0, 1, 8), np.linspace(0, 1, 8))

pset = ParticleSet(fieldset, pclass=SampleParticle, lon=xv.flatten(), lat=yv.flatten())
pset.execute(SampleP, endtime=1, dt=1)
pset
```

Outputs:

```
P[320](lon=0.000000, lat=0.000000, depth=0.000000, p=0.000000, time=1.000000)
P[321](lon=0.142857, lat=0.000000, depth=0.000000, p=0.000000, time=1.000000)
...
```
i.e. `p` is 0 at `t=1` while this corresponds to its value at `t=0` but not `t=1`.

**Comment by erikvansebille** [comment]
Hmm, this is a very interesting point; thanks for raising it, @apatlpo. Indeed, the order of operations in Parcels is 

Kernel evaluations -> updating time -> writing output. That has so far gone well, but in this case leads to unexpected behaviour. If you give `endtime=2` in the `pset.execute()`, you will see that `p=1` (but `t=2`)

What use-case do you have for this. Is there a way to circumvent this issue?

**Comment by erikvansebille** [comment]
One fairly simple fix (hack?) could be to change your sampling kernel to
```python
def SampleP(particle, fieldset, time):
    particle.p = fieldset.P[time+particle.dt, particle.depth, particle.lat, particle.lon]
```
(i.e. sample at time `time+particle.dt`)

**Comment by apatlpo** [comment]
Thanks for the feedback and fixes @erikvansebille !

I did not try to figure out whether the sampled field corresponds to `p[t-dt, x(t)]` or `p[t-dt, x(t-dt)]`.
Would you know for both the original example and the fix you suggest (`particle.p = fieldset.P[time+particle.dt, ...`)?

I run into this while comparing offline interpolations of original fields with online ones.
When you are interesting in the signature and dynamics of high frequency processes (e.g. tides, internal waves) and the temporal resolution is not that small compared to these processes' timescales, such offset may matter.

**Comment by apatlpo** [comment]
> What if you just interpret all outputs as t-dt (i.e. subtract dt from each output when you analyse them); do the offline interpolations then compare well with the online ones?

Yes, that's what I had in mind.
If understand correctly your last post, the sampled output corresponds to `p[t-dt, x(t-dt)]` and NOT `p[t-dt, x(t)]`.  Could you please confirm this point?

**Comment by erikvansebille** [comment]
Yes, correct, if you use the kernel 
```python
def SampleP(particle, fieldset, time):
    particle.p = fieldset.P[time, particle.depth, particle.lat, particle.lon]
```
then the output corresponds to `p[t-dt, x(t-dt)]`. 

It's just that `t` is updated (with `dt`) _after_ the sampling but before the output, so there can be a `dt` mismatch

**Comment by apatlpo** [comment]
One last thing (apologizes): 
could also please confirm that, on the other hand, the position (lon, lat) is indeed that at time `t` (an NOT `t-dt`)?

**Comment by erikvansebille** [comment]
Well, good question. I now realise that this depends on which order you concatenate your sampling and advection kernel. 

They are executed in order, so if your kernels are `pset.kernel(AdvectionRK4)+pset.kernel(SampleP)`, then the positions are _first_ updated and then the sampling is done on the locations at `t`. 

If you use `pset.kernel(SampleP)+pset.kernel(AdvectionRK4)` then the sampling is done first so the sampling location is at `t-dt`

**Comment by apatlpo** [comment]
And I have been using the later order so far which would explain why there is an offset.

If I understand correctly, the fix is simple then and requires to set the former kernel order.
I'll give it a shot with the example above to verify first.

If you have an idea for where such information should be found and would like a contribution, I can give it a shot.

**Comment by erikvansebille** [comment]
Very keen to hear if switching the order fixes your issue! 

If so, we should indeed find a way to communicate this subtlety better. Do you have any idea where you would have liked to read about how the order of concatenating Kernels matters? Which notebook tutorials etc did you read before you started using Parcels?

**Comment by apatlpo** [comment]
Unless I got something wrong, switching the order does not fix the issue unfortunately:
```
pset = ParticleSet(fieldset, pclass=SampleParticle, lon=xv.flatten(), lat=yv.flatten())
kernels = pset.Kernel(AdvectionRK4) + pset.Kernel(SampleP)
pset.execute(kernels, endtime=1, dt=1)
pset[0]
```
and
```
pset = ParticleSet(fieldset, pclass=SampleParticle, lon=xv.flatten(), lat=yv.flatten())
#kernels = pset.Kernel(SampleP) + AdvectionRK4
kernels = pset.Kernel(SampleP) + pset.Kernel(AdvectionRK4)
pset.execute(kernels, endtime=1, dt=1)
pset[0]
```
both lead to something like:
```
P[896](lon=0.000000, lat=0.000000, depth=0.000000, p=0.000000, time=1.000000)
```

**Comment by apatlpo** [comment]
fyi, double checking the my previous post.
I may have gotten something wrong

**Comment by apatlpo** [comment]
Allright, I did some experiments see following [gist](https://gist.github.com/apatlpo/26593913378c1e041cde08a722b3d28c).

It seems the correct solution is to do something like:

```
def SamplePoff(particle, fieldset, time):
    # pay attention to the dt offset
    particle.p = fieldset.P[time+particle.dt, particle.depth, particle.lat, particle.lon]

kernels = AdvectionRK4 + pset.Kernel(SamplePoff)
pset.execute(kernels, ...)
```

where the order in the sum of kernels matters (the sampling uses inconsistent position and time otherwise).
Do you believe this is a robust solution?

**Comment by erikvansebille** [comment]
Thanks for this [gist](https://gist.github.com/apatlpo/26593913378c1e041cde08a722b3d28c) @apatlpo! Makes total sense to me. If you first run `AdvectionRK4`, the particle positions _have_ been updated, but the time hasn't. So if you then sample the field P at t+dt, you will get your expected results because before the run finishes, `time += dt`, so then all variables are at the new time

Note that it may be good to check if this behaviour is the same for writing output to file during the `pset.exec()`. I'm not entirely sure whether `time` gets updated before or after writing to file...

**Comment by apatlpo** [comment]
Just updated the gist with netcdf outputs.
No bad surprise, the content of netcdf files follows the data held in the pset object.

Coming back to where this information should be found.
I based my tests on the [interpolation tutorial](https://nbviewer.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_interpolation.ipynb) which could be extended to include the above mentioned information.

**Comment by erikvansebille** [comment]
Thanks @apatlpo, I have now added a section to the interpolation tutorial, see #1127 to highlight this important(!) subtlety. Let me know if you think this can be improved

**Comment by apatlpo** [comment]
Yes @erikvansebille this will certainly do it.

The explicit mention to me is not necessary and I would drop it.
But pointing to the issue is certainly useful.

Thanks again for your reactivity in all of this @erikvansebille your pointers saved me a lot of time !

**Comment by erikvansebille** [comment]
Solved in #1127
